### PR TITLE
Source available metrics

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Release Prometheus metrics under the [Source Available License](https://docs.mattermost.com/about/faq-mattermost-source-available-license.html):

> A source available license gives access to source code, but places restrictions on its use. The Mattermost Source Available License allows free-of-charge and unrestricted use of the source code in development and testing environments, but requires a valid Mattermost Enterprise Edition License in a production environment.

By default, the new `github.com/mattermost/mattermost/server/v8/enterprise` package is effectively only compiled into enterprise builds (`BUILD_ENTERPRISE=true`), and thus does not ship with the team edition (preserving the fully open source philosophy there). But in development environments with a `model.BuildNumber == "dev"` (automatically set via the `Makefile`) the `:8067/metrics` endpoint is made available (if the config.json `MetricSettings.Enable == true`). No access to the closed source enterprise repository is required to develop against or contribute to this portion of our enterprise product.

As a side effect of leveraging build tags (`enterprise` and `sourceavailable`), some of the complexity with `channels/imports/imports.go` is avoided altogether.

#### Ticket Link
None.

#### Release Note
```release-note
Prometheus metrics are now available under the Source Available License.
```

### CODE REVIEW

File: server/.gitignore
Comment (lieut-data) [inline server/.gitignore:25]: I've left the `rm -f` in `make clean`, but I note that transitioning across this PR might be slightly painful. Thoughts on doing a `rm -f` whenever the `Makefile` is invoked if the file is detected?

File: server/Makefile
Comment (lieut-data) [inline server/Makefile:36]: Use this to accumulate build tags as we decide to set them.

File: server/Makefile
Comment (lieut-data) [inline server/Makefile:140]: Use this since Go doesn't support `tags tag1 -tags tag2 ...`.

File: server/Makefile
Comment (lieut-data) [inline server/Makefile:182]: No longer needed, since the equivalent logic is handled by build tags.

File: server/Makefile
Comment (lieut-data) [inline server/Makefile:774]: This is temporary until https://github.com/mattermost/mattermost-govet/pull/38 is reviewed.

File: server/enterprise/metrics/metrics.go
Comment (lieut-data) [inline server/enterprise/metrics/metrics.go:400]: One minor change here not to assume `m.Platform.Cluster()` since we can now run this in development environments without that enterprise feature.

File: server/Makefile
Comment (agnivade) [inline server/Makefile:142]: I tested with `-tags ""` and it works fine. So I don't think we need this. Rather simply in the Go command, we can write `-tags "$(BUILD_TAGS)"`

File: server/build/release.mk
Comment (agnivade) [inline server/build/release.mk:1]: Same comment as earlier. Would prefer to embed this at the call site itself.

File: server/channels/app/platform/service.go
Comment (agnivade) [inline server/channels/app/platform/service.go:329]: I am slightly confused by this. We should be restarting the metrics if a new license gets added having metrics feature right? But in this case, it looks like any change from old to new means we don't do anything.

File: server/Makefile
Comment (lieut-data) [reply server/Makefile:142]: Ooh, that's much more elegant -- thank you!

File: server/channels/app/platform/service.go
Comment (lieut-data) [reply server/channels/app/platform/service.go]: You're right! I'm not sure how I missed this during testing. Let me explore the e2e suite to see if this is something I can automatically test.

File: server/Makefile
Comment (lieut-data) [inline server/Makefile:96]: Any objection to this, @agnivade & @crspeller? I literally ran into the thing I was worried about today, scratched my head, then realized my mistake. I'd prefer not to have this repeated over and over again as people upgrade, or later as they switch between branches.

Comment (crspeller) [reply server/Makefile:96]: No objection. I would prefer not to have to remember to delete this file constantly.

File: server/channels/app/platform/service.go
Comment (crspeller) [inline server/channels/app/platform/service.go:329]: This is a mind bender. Are we sure this is correct? It appears to call `RestartMetrics` if the license state has changed, but there appears to be no actual license check here? Is it handled somewhere else?

**Review by crspeller — APPROVED** [review]
This is great!

File: server/Makefile
Comment (agnivade) [reply server/Makefile:96]: Fine by me!

**Comment by lieut-data** [comment]
@esarafianou, uncertain if you think this pull request requires security review, but wanted to give you an opportunity to take a peek before we merge.

File: server/channels/app/platform/service.go
Comment (lieut-data) [reply server/channels/app/platform/service.go:329]: Agree it's convoluted. We do the actual check in 

https://github.com/mattermost/mattermost/blob/8a08714ae15c1361f1a0cb3deab4b327be2d5ead/server/enterprise/metrics/metrics.go#L954-L962

**Comment by agnivade** [comment]
@lieut-data - Feel free to request a review when it's ready.

**Comment by lieut-data** [comment]
> @lieut-data - Feel free to request a review when it's ready.

Oops, I see now you only commented and didn't actually approve. My bad! It should have been ready for review since I added you and @crspeller. (I just keep fixing merge conflicts with upstream.)

**Review by agnivade — APPROVED** [review]
Nice work!

**Comment by lieut-data** [comment]
@saturninoabril, this /should/ have no end user impact, with the existing and newly added e2e tests confirming same. I've also tested it locally to verify metrics functionality, so far without issues. Thoughts on any additional testing that we should require before merging?

File: e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js
Comment (saturninoabril) [inline e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js:11]: `MetricsSettings` is not allowed to be changed in cloud-like test server and so we'll not test this in such environment.

```suggestion
// Group: @channels @enterprise @metrics @not_cloud
```

File: e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js
Comment (saturninoabril) [inline e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js:15]:
```suggestion
    before(() => {
        cy.shouldNotRunOnCloudEdition();
        cy.apiUpdateConfig({
```

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks Jesse! Overall the test looks good to me. I don't think there's more to add given the changes made here. I can see this as a good start to campaign for more metrics test. I'll add it on my list.

I left few comments, mostly about how the test will run in CI. It's currently failing btw but probably due to 8067 not open in CI.
Screenshot 2023-10-19 at 3 58 36 PM

**Comment by esarafianou** [comment]
@lieut-data I skimmed through the PR and I didn't see any issues but didn't thoroughly review it. Feel free to remove my as a reviewer.

File: e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js
Comment (lieut-data) [reply e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js:26]: Ah, yes, if `ListenAddress` is just `:8067`, this fails. (On my localhost, I make `localhost:8067` explicit. Will change to make matching the host optional:

```suggestion
            baseURL.port = config.MetricsSettings.ListenAddress.replace(/^.*:/, '');
```

File: e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js
Comment (lieut-data) [reply e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js:52]: Understood, thank you for clarifying the nuances! I'll rework this as per the current limitations. I wonder if we could leverage the service environment to "safely" share a developer license that we don't care about risking exposure? That would sidestep the issues above with the slight caveat that we'd be testing "non-production" in a very small number of situations (all of which actually apply perfectly to a testing environment, e.g. not hitting Rudder). A smoketest that used a production license could validate the rest.

cc @marianunez for thoughts?

Comment (lieut-data) [reply e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js]: I can confirm from an `/e2e-test` run that we're already using the `test` service environment:
CleanShot 2023-10-19 at 14 46 52@2x

@saturninoabril, does this mean we can dispense with concerns around leaking license and even go so far as to check in a set of licenses for various situations that we can upload and delete at will?

Comment (saturninoabril) [reply e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js]: If all are good with test licenses potentially leaking and no issue with it, then it's great as it will unblock test like this and others as well. I'll go ahead and enable in the CI.
Thanks @lieut-data and @marianunez for confirming!

Comment (lieut-data) [reply e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js]: cc @iyampaul & @esarafianou, as per the above, our plan is to check in some test licenses that unlock a more holistic testing strategy, knowing full well that these may "leak". The licenses are only useful if `ServiceEnvironment == test`, but I admit there are few "consequences" for forcing such a service environment. That being said, I note this is "the GitLab way". Any thoughts or concerns?

**Review by saturninoabril — APPROVED** [review]
Thanks Jesse! All good to me. Other is non-blocking.

File: e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js
Comment (iyampaul) [reply e2e-tests/cypress/tests/integration/channels/enterprise/metrics/license_spec.js]: I'm not overly worried about this; the people using a leaked token for prod will be the same ones who will try to find other means of circumventing our licensing/auth method anyways. Regardless of them being able to unlock EE features they still won't get support from us if they need it, and will need to re-apply their "crack" each update.

**Review by iyampaul — APPROVED** [review]

**Comment by lieut-data** [comment]
The only outstanding task on this pull request is to fix the issue of metrics reloading/unloading on a license swap that makes the e2e test invalid. (It would be fine on a restart, but hard to test when fully online.) Once that's done, I'll aim to merge.

**Comment by lieut-data** [comment]
@saturninoabril, I thought I had root caused why the e2e test fails in CI, but I'm now running into another issue that relates to the admin user not being allowed to remove the license. It seems the admin by this point in the suite is a "restricted system admin" instead of a full admin:
CleanShot 2023-11-14 at 15 22 16@2x

any idea what might be causing that from running the full suite?

**Comment by saturninoabril** [comment]
@lieut-data It might be due to the one that changes the setting to `ExperimentalSettings.RestrictSystemAdmin` when running against self-hosted server - [authentication/password_settings_spec.js](https://github.com/mattermost/mattermost/blob/master/e2e-tests/cypress/tests/integration/channels/system_console/authentication/password_settings_spec.js#L17). I've updated the filter to sort it last - https://git.internal.mattermost.com/qa/cypress-ui-automation/-/commit/cd56dd5f961a3e11473b8ee400fffef9ffeb9148

**Comment by saturninoabril** [comment]
@lieut-data E2E tests failed due to image being expected at `mattermostdevelopment/mattermost-enterprise-edition`. Could you fix the failing build then we can try E2E again?

**Comment by lieut-data** [comment]
@saturninoabril, I think I've figured out why this fails during e2e but not locally run. The logs for the Mattermost instance running this spec have these entries (the only `DELETE /api/v4/license` in that run):
```
{"timestamp":"2023-12-01 19:18:42.432 Z","level":"info","msg":"Remove license.","caller":"platform/license.go:265","id":"ActiveLicenseId"}
{"timestamp":"2023-12-01 19:18:42.432 Z","level":"error","msg":"Elasticsearch.start: Elasticsearch is already stopped.","caller":"platform/searchengine.go:68"}
{"timestamp":"2023-12-01 19:18:42.433 Z","level":"debug","msg":"Skipping triggering Elasticsearch channel index fix job as build is not Enterprise ready","caller":"app/migrations.go:567"}
{"timestamp":"2023-12-01 19:18:42.436 Z","level":"info","msg":"Purging all caches","caller":"platform/cluster_handlers.go:146"}
{"timestamp":"2023-12-01 19:18:42.438 Z","level":"info","msg":"License key from ENV is valid, unlocking enterprise features.","caller":"platform/license.go:78"}
{"timestamp":"2023-12-01 19:18:42.439 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:172","method":"DELETE","url":"/api/v4/license","request_id":"tczcnswzh78ednumgk4doknp9w","user_id":"bp1niodhcjfmuyds5am5o895jc","status_code":"200"}
```

Notably, given `License key from ENV is valid, unlocking enterprise features.`, it sounds like we set the license for tests via `MM_LICENSE`, and that this effectively prevents deleting the license at all. I can reproduce this locally when `MM_LICENSE` is used:

```
$ http :8065/api/v4/license/client format==old | jq '.IsLicensed'
"true"
$ http --print h DELETE :8065/api/v4/license | head -n1
HTTP/1.1 200 OK
$ http :8065/api/v4/license/client format==old | jq '.IsLicensed'
"true"
```

by contrast, if `MM_LICENSE` is not in use and I just upload a license manually:
```
$ http :8065/api/v4/license/client format==old | jq '.IsLicensed'
"true"
$ http --print h DELETE :8065/api/v4/license | head -n1
HTTP/1.1 200 OK
$ http :8065/api/v4/license/client format==old | jq '.IsLicensed'
"false"
```

I think this means that all tests currently using `cy.apiDeleteLicense` are probably invalid, in that they don't actually test what happens when the server isn't licensed. Thoughts on how to proceed from here?

**Comment by saturninoabril** [comment]
Thanks @lieut-data for investigating! Yes, we're setting up the license via env variable in CI. Feel free to merge. We'll make a follow up adjustment in test setup accordingly -- maybe either change how it loads the license or totally separate TE-related tests.

For server using `MM_LICENSE`, I thought it can be deleted in the past but I don't remember. Do you think the current behavior is totally fine and is as expected?

**Review by saturninoabril — APPROVED** [review]
Thanks Jesse!

**Comment by lieut-data** [comment]
> For server using `MM_LICENSE`, I thought it can be deleted in the past but I don't remember. Do you think the current behavior is totally fine and is as expected?

If `MM_LICENSE` is set, it's probably "expected" that one can't really delete the license. Ideally, the application however at least fail the deletion request as though it was the read-only setting it is, but probably not urgent.

**Comment by cwarnermm** [comment]
@lieut-data - What do Mattermost admins or end users need to know about this PR via the product documentation?

**Comment by lieut-data** [comment]
@cwarnermm, no documentation changes required for administrators or end users. This is entirely developer facing.

**Pull request merged**
